feat(vote): add vote component

dancormier opened the pull request:
[SPARK-103](https://stackoverflow.atlassian.net/browse/SPARK-103)

---

[Figma](https://www.figma.com/design/do4Ug0Yws8xCfRjHe9cJfZ/Project-SHINE---Product-UI?node-id=617-18830&p=f&m=dev)
[Docs (Stacks Classic)](https://deploy-preview-2062--stacks.netlify.app/product/components/vote/)
[Storybook (Stacks Svelte)](https://deploy-preview-2062--stacks-svelte.netlify.app/?path=/docs/components-vote--docs)

---

This PR adds a _vote_ component to Stacks Classic and Stacks Svelte.

> [!NOTE]
> At the time of writing, downvote icons are not yet design so are instead shown using the placeholder icons. Once the downvote icons are updated in the stacks-icons packages, the correct icons should show.

<details>
<summary>Screenshots</summary>

### Base
#### Light
<img width="441" height="160" alt="image" src="https://github.com/user-attachments/assets/bad3bc9e-fb37-4362-b2bb-cc5eefaaecbf" />

#### Dark
<img width="439" height="156" alt="image" src="https://github.com/user-attachments/assets/4935c885-420c-4551-99e0-d38797e44e83" />

#### Light HC
<img width="442" height="162" alt="image" src="https://github.com/user-attachments/assets/c7441577-c960-404a-9cac-d80084f11f0c" />

#### Dark HC
<img width="440" height="157" alt="image" src="https://github.com/user-attachments/assets/9bcee38e-f737-4bc0-8589-ae9b93aff586" />

### Expanded
<img width="75" height="139" alt="image" src="https://github.com/user-attachments/assets/28edfebd-9620-40a5-a612-492878243999" />

### Horizontal
<img width="403" height="89" alt="image" src="https://github.com/user-attachments/assets/c97953cc-babd-45a7-9528-8b27d3ae940a" />

### Voted
<img width="736" height="169" alt="image" src="https://github.com/user-attachments/assets/9fba4abf-0a24-4eaa-b9e5-f21e5e7c3483" />

</details>

dependabot[bot] committed:
build(deps-dev): bump stylelint-config-standard from 38.0.0 to 39.0.1 (#2004)

Bumps [stylelint-config-standard](https://github.com/stylelint/stylelint-config-standard) from 38.0.0 to 39.0.1.
- [Release notes](https://github.com/stylelint/stylelint-config-standard/releases)
- [Changelog](https://github.com/stylelint/stylelint-config-standard/blob/main/CHANGELOG.md)
- [Commits](https://github.com/stylelint/stylelint-config-standard/compare/38.0.0...39.0.1)

---
updated-dependencies:
- dependency-name: stylelint-config-standard
  dependency-version: 39.0.1
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dancormier committed:
feat(vote): add vote (btn(s)) component

dancormier committed:
Rename to just "vote", better styles

dancormier committed:
Add expanded modifier, positive/negative count styles

dancormier committed:
Update docs

dancormier committed:
Add basic tests

dancormier committed:
Better, but not perfect, vote icons

dancormier committed:
Add basic Svelte component

dancormier committed:
improve svelte component, update icons, add formatNumber, and more!

dancormier committed:
A lil renaming and cleanup

dancormier committed:
Add a11y, visual tests

dancormier committed:
Add svelte test and cleanup

dancormier committed:
Doc updates

dancormier committed:
More docs updates

dancormier committed:
Merge branch 'beta' into SPARK-103/vote-btn-component

dancormier committed:
Use moduleResolution: "bundler" in Stacks Classic

dancormier committed:
Cleanup

dancormier committed:
Improve stacks classic tests

dancormier committed:
Add i18nExpand, i18nExpanded

dancormier committed:
Update svelte test

dancormier committed:
Svelte component tweaks

dancormier committed:
changeset

dancormier committed:
Update visual test

dancormier committed:
Add visual regression test images

dancormier committed:
formatting, tests updates

dancormier committed:
Fix a test

dancormier committed:
formatting!!!!

mukunku commented:
Not sure if this is ready for review but noticed the downvote button looks weird: https://deploy-preview-2062--stacks-svelte.netlify.app/?path=/docs/components-vote--docs
<img width="215" height="395" alt="{D90C11DB-CAD4-4428-865E-9ACC1050C816}" src="https://github.com/user-attachments/assets/a51ed182-cd7b-4bc6-9251-6406e1b0b65c" />

mukunku commented on packages/stacks-docs/product/components/vote.html:
```diff
@@ -0,0 +1,242 @@
+
+    {% header "h3", "Expanded" %}
+    <p class="stacks-copy">
+        Include the <code class="stacks-code">.s-vote__expanded</code> modifier to show upvote and downvote counts in addition to the total vote count. This modifier hides <code class="stacks-code">.s-vote--total</code> and shows <code class="stacks-code">.s-vote--upvotes</code> and <code class="stacks-code">.s-vote--downvotes</code> instead.
```
Should this be "instead of" instead of "in addition to"? Seems like the very next sentence contradicts the first one.

mukunku commented on packages/stacks-docs/product/components/vote.html:
```diff
@@ -0,0 +1,242 @@
+
+    {% header "h3", "Horizontal" %}
+    <p class="stacks-copy">
+        Apply the <code class="stacks-code">.s-vote__horizontal</code> modifier to arrange the vote buttons and counts in a horizontal layout. This layout does not officially support expanded vote count. This configuration is best suited scenarios such as comment voting, where a more compact design is preferred.
```
small nit: missing "for" 
```suggestion
        Apply the <code class="stacks-code">.s-vote__horizontal</code> modifier to arrange the vote buttons and counts in a horizontal layout. This layout does not officially support expanded vote count. This configuration is best suited for scenarios such as comment voting, where a more compact design is preferred.
```

mukunku reviewed (commented):
Changes are looking great. Noticed one thing: the selected state for horizontal layout doesn't match the one for comments in designs. Similarly, you can't click on the whole thing to upvote which I'm guessing isn't the intention 🤔 

<img width="367" height="158" alt="{279EFF0C-0FD5-4DC7-8971-7C1DA5E9CDFF}" src="https://github.com/user-attachments/assets/58264da7-8106-49ce-82e3-0d43eea93fcb" />
<img width="1011" height="420" alt="{26D5D4B6-493D-4237-9F94-0AD33ACBB63A}" src="https://github.com/user-attachments/assets/670deeae-79d2-4586-99d1-2565dc7bc6d3" />

Do we need an `upvote-only` mode or something maybe?

CGuindon commented:
Downvote icons are available now (the vote icon but direction set to down)

CGuindon commented:
[Confirming](https://github.com/StackExchange/Stacks/pull/2062#pullrequestreview-3572999327) we don't have a use case currently where we have a horizontal version with both up and downvotes. We only have the horizontal single vote so wondering if we should just **not** show the version with 2 votes on horizontal (people shouldn't use this). 

I'm ok keeping the version you have though where both up/down are possible and the focus/hover stays on one side. I think it still works fine even with just the upvote and it would save us time later if product decides to add in downvotes to comments/replies.

CGuindon reviewed (changes requested):
New downvote icon needed (overall looks good though!)

dancormier committed:
add missing "for"

Co-authored-by: Sal <sbatirbek@stackoverflow.com>

dancormier committed:
Match horizontal vote mocks

dancormier committed:
Merge branch 'beta' into SPARK-103/vote-btn-component

dancormier committed:
Address reviews

dancormier commented:
> the selected state for horizontal layout doesn't match the one for comments in designs

Nice catch @mukunku! I built it with this in mind originally but completely spaced when making the examples and Svelte component. Fixed!

---

> I'm ok keeping the version you have though where both up/down are possible and the focus/hover stays on one side. I think it still works fine even with just the upvote and it would save us time later if product decides to add in downvotes to comments/replies.

@CGuindon Sal pointed out that I missed the configuration in the Figma. I've updated the horizontal orientation to match what's in Figma where the element is just one upvote button that includes the vote count. If/when we want to allow for downvotes on the horizontal orientation, we'll be able to do that pretty trivially.

> Downvote icons are available now (the vote icon but direction set to down)

The standard (20x20) downvote icon is available but we need the 16px ones for this component. We can still merge without them and once they're published and the icons library dependency is updated, this component will display them with no further changes besides regenerating the test images.

mukunku reviewed (approved):
LGTM!

giamir reviewed (approved):
Great work @dancormier. I really like this component. 

The only thing I noticed is that we automatically update the total number but not the upvotes and downvotes numbers. 

<img width="105" height="152" alt="Screenshot 2025-12-17 at 10 46 47" src="https://github.com/user-attachments/assets/1842df1f-cad4-4895-84be-407cb18773d3" />

I am actually unsure if updating the different counts is something we should do in the design system component directly or rather have the consumers take care of it.  We don't have control over what the server returns and how the API looks like and we risk to show an incorrect value to the user (imagine a call is done to the server to register an upvote and the server come back with a different score because in the meantime somebody else has also casted a vote, etc...).

That said I am happy to merge it as is and maybe have those conversations with the first team implementing the component. 

A small nit. the prop we call `total` in the component is called `score` in our APIs (check for [example](https://sem1.dev.stackenterprise.co/api/v3/swagger/index.html) the schema of the response for `questions/id`). It would be nice probably to stick to score also here but total is ok too. 🙂

CGuindon commented:
The Svelte component was built to update the count and wait for a promise to resolve to confirm the vote count was updated on the server-side. It makes sense to do the same thing for the separate counts. 

[Slack thread](https://stackexchange.slack.com/archives/C097YESGSMB/p1765987695613159)
